pkg/binding/selectors.go

```diff
@@ -117,17 +77,89 @@ func (c *Controller) updateDecisions(obj runtime.Object) error {
 			}
 
 			return fmt.Errorf("failed to update resolution for bindingpolicy %s for object %v: %v",
-				bindingpolicy.GetName(), util.RefToRuntimeObj(obj), err)
+				bindingPolicy.GetName(), util.RefToRuntimeObj(obj), err)
 		}
 
 		if resolutionUpdated {
 			// enqueue binding to be synced since an object was added to its bindingpolicy's resolution
 			c.logger.V(4).Info("enqueued Binding for syncing due to a noting of an "+
-				"object in its resolution", "binding", bindingpolicy.GetName(),
+				"object in its resolution", "binding", bindingPolicy.GetName(),
 				"object", util.RefToRuntimeObj(obj))
-			c.enqueueBinding(bindingpolicy.GetName())
+			c.enqueueBinding(bindingPolicy.GetName())
+		}
+
+		// make sure object has singleton status if needed
+		if !isBeingDeleted(obj) &&
+			c.bindingPolicyResolver.ResolutionRequiresSingletonReportedState(bindingPolicy.GetName()) {
+			if err := c.handleSingletonLabel(ctx, obj, bindingPolicy.GetName()); err != nil {
+				return fmt.Errorf("failed to add singleton label to object: %w", err)
+			}
+
+			isSelectedBySingletonBinding = true
+		}
+	}
+
+	// if the binding-policies matching cycles end and the object is not selected by a singleton binding,
+	// we need to remove the singleton label from the object if it exists.
+	// NOTE that this takes care of the case where the object was previously selected by a singleton binding
+	// and is no longer selected by any binding.
+	if !isBeingDeleted(obj) && !isSelectedBySingletonBinding {
+		if err := c.handleSingletonLabel(ctx, obj, emptyBindingPolicyName); err != nil {
+			return fmt.Errorf("failed to remove singleton label from object: %w", err)
 		}
 	}
 
 	return nil
 }
+
+// handleSingletonLabel adds or removes the singleton label from the object. If a bindingPolicyName is provided,
+// the singleton label is added to the object. If the bindingPolicyName is empty, the singleton label is removed.
+func (c *Controller) handleSingletonLabel(ctx context.Context, obj runtime.Object, bindingPolicyName string) error {
+	unstructuredObj, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		return fmt.Errorf("failed to convert runtime.Object to unstructured.Unstructured")
+	}
+
+	labels := unstructuredObj.GetLabels()
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+
+	_, found := labels[util.BindingPolicyLabelSingletonStatusKey]
+
+	if bindingPolicyName == emptyBindingPolicyName {
+		if !found {
+			return nil
+		}
+		delete(labels, util.BindingPolicyLabelSingletonStatusKey)
```

Comment:
Is it safe to side-effect the given `obj`? I see nothing in the comments, naming, or type system that answers the question. This is bad.